cosi/protocol/protocol.go

```diff
@@ -0,0 +1,346 @@
+package protocol
```

Comment:
I stopped reviewing here, and will pick it back up tomorrow morning.

Reply:
Thanks. My latest commits and comments hopefully address your review so far.